srml/support/procedural/src/storage/mod.rs

```diff
@@ -14,218 +14,313 @@
+		);
+
+		Self {
+			hidden_crate: def.hidden_crate,
+			visibility: def.visibility,
+			store_trait: def.store_trait,
+			module_name: def.module_name,
+			module_runtime_generic: def.module_runtime_generic,
+			module_runtime_trait: def.module_runtime_trait,
+			module_instance: def.module_instance,
+			where_clause: def.where_clause,
+			extra_genesis_build: def.extra_genesis_build,
+			extra_genesis_config_lines: def.extra_genesis_config_lines,
+			crate_name: def.crate_name,
+			storage_lines,
+			module_struct,
+			module_impl,
+			optional_instance,
+			optional_instance_bound,
+			optional_instance_bound_optional_default,
+		}
+	}
 }
 
-#[derive(Parse, ToTokens, Debug)]
-enum AddExtraGenesisLineEnum {
-	AddExtraGenesisLine(AddExtraGenesisLine),
-	AddExtraGenesisBuild(DeclStorageBuild),
+pub struct ModuleInstanceDef {
+	instance_generic: syn::Ident,
+	instance_trait: syn::Ident,
+	instance_default: Option<syn::Ident>,
 }
 
-#[derive(Parse, ToTokens, Debug)]
-struct AddExtraGenesisLine {
-	pub attrs: ext::OuterAttributes,
-	pub config_keyword: keyword::config,
-	pub extra_field: ext::Parens<Ident>,
-	pub coldot_token: Token![:],
-	pub extra_type: syn::Type,
-	pub default_value: ext::Opt<DeclStorageDefault>,
+pub struct StorageLineDef {
+	attrs: Vec<syn::Attribute>,
+	visibility: syn::Visibility,
+	name: syn::Ident,
+	getter: Option<syn::Ident>,
+	config: Option<syn::Ident>,
+	build: Option<syn::Expr>,
+	default_value: Option<syn::Expr>,
+	storage_type: StorageLineTypeDef,
 }
 
-#[derive(Parse, ToTokens, Debug)]
-struct DeclStorageLine {
-	// attrs (main use case is doc)
-	pub attrs: ext::OuterAttributes,
-	// visibility (no need to make optional
-	pub visibility: syn::Visibility,
-	// name
-	pub name: Ident,
-	pub getter: ext::Opt<DeclStorageGetter>,
-	pub config: ext::Opt<DeclStorageConfig>,
-	pub build: ext::Opt<DeclStorageBuild>,
-	pub coldot_token: Token![:],
-	pub storage_type: DeclStorageType,
-	pub default_value: ext::Opt<DeclStorageDefault>,
+pub struct StorageLineDefExt {
+	#[allow(unused)]
+	attrs: Vec<syn::Attribute>,
+	visibility: syn::Visibility,
+	name: syn::Ident,
+	getter: Option<syn::Ident>,
+	config: Option<syn::Ident>,
+	build: Option<syn::Expr>,
+	default_value: Option<syn::Expr>,
+	storage_type: StorageLineTypeDef,
+	doc_attrs: Vec<syn::Meta>,
+	/// Either the type stored in storage or wrapped in an Option.
+	query_type: syn::Type,
+	/// The type stored in storage.
+	value_type: syn::Type,
+	storage_struct: proc_macro2::TokenStream,
+	optional_storage_runtime_comma: Option<proc_macro2::TokenStream>,
+	optional_storage_runtime_bound_comma: Option<proc_macro2::TokenStream>,
+	optional_storage_where_clause: Option<proc_macro2::TokenStream>,
+	storage_trait: proc_macro2::TokenStream,
+	storage_generator_trait: proc_macro2::TokenStream,
+	is_generic: bool,
+	is_option: bool,
 }
 
+impl StorageLineDefExt {
+	fn from_def(storage_def: StorageLineDef, def: &DeclStorageDef) -> Self {
```

Comment:
Maybe make it consistent with `impl From<DeclStorageDef> for DeclStorageDefExt` by implementing sth like 
`impl From<(StorageLineDef, &DeclStorageDef)> for StorageLineDef `

Reply:
hmm I usually don't very like to use From trait for tuple types like this. this looks more like a function signature than a type by itself.